### ISSUE TITLE
Factory method makes InMemorySubscriptionStore disapear

### ISSUE THREAD

**Issue by spawluk** [rebus-org/Rebus#675]
As coded in:

https://github.com/rebus-org/Rebus/blob/ceebbabe951e1ac9151a2cd9828f9fafb7d61782/Rebus/Persistence/InMem/InMemorySubscriptionStorageExtensions.cs#L33

and using configuration code:
```
Configure.With(adapter)
        .Transport(x => x.UseInMemoryTransport(new InMemNetwork(), "bus"))
        .Subscriptions(x => x.StoreInMemory())
        .Start();
```

every time when Rebus tries to subscribe to message or tries to receive message, new instance of `InMemorySubscriberStore` is created. It leads to situation when configured bus cannot receive published message from topic, because on message receive new, empty instance of subscription store is created.

Code that actually works:

```
Configure.With(adapter)
        .Transport(x => x.UseInMemoryTransport(new InMemNetwork(), "bus"))
        .Subscriptions(x => x.StoreInMemory(new InMemorySubscriberStore()))
        .Start();
```

**Comment by mookid8000** [comment]
> every time when Rebus tries to subscribe to message or tries to receive message, new instance of `InMemorySubscriberStore` is created

I don't think that's correct 😄 

If you experience a difference in behavior, it might be caused by timing isses and the fact that the in-mem subscription storage works in the _decentralized mode_ by default.

_Decentralized mode_ means that publishers keep their own subscription storages private, and then subscribers subscribe/unsubscribe by sending messages to the publisher.

You can read more about it [here](https://github.com/rebus-org/Rebus/wiki/Pub-sub-messaging#transports-without-native-pubsub-support)

But I must admit I cannot readily see how you can come to the conclusion that the in-mem store is being newed up all the time – can you tell me more about how you came to that conclusion?

**Comment by spawluk** [comment]
Hello and thanks for the response.

My conclusion was based on fast debugging of RebusBus instance and its local field `_subscriptionStore`.

I was mislead by fact that this store is empty, even after subscribing, and my test message was not handled by handler. Even log message `Sending "System.Object" -> "<no destinations>"` made me think that no subscription is saved.

While writing this response I could not recreate this issue, because I switched to `StructureMapContainerAdapter`. Both approaches (centralized and decentrilized) work using this activator.

But after changing Activator to built-in, and waiting for subscription event to be handled I cannot receive an event.

Then I checked Interface of built-in activator and... Bummer! I didn't register any of event handlers!

So finishing my strange case code:

```
var activator = new BuiltinHandlerActivator();

activator.Handle<object>(o =>
  {
    Log.Information("Handled");
    return Task.CompletedTask;
  });

var bus = Configure.With(activator)
  .Transport(x => x.UseInMemoryTransport(new InMemNetwork(true), "bus"))
  .Subscriptions(x => x.StoreInMemory())
  .Routing(x => x.TypeBased().Map<object>("bus"))
  .Start();

Task.WaitAll(bus.Subscribe<object>());
Task.WaitAll(bus.Publish(new object()));
```

Works like magic.
